### ISSUE TITLE
websocket error: TypeError: server.listeners is not a function

### ISSUE THREAD

**Issue by warriorBrian** [nestjs/nest#6992]
## Bug Report
I'm creating a basic websocket example using nest.js, but it's not working properly.
## Current behavior

I'm creating a basic WebSocket connection based on the example. According to the example, the default port is used, but the console will report an error:

```shell
(node:98373) UnhandledPromiseRejectionWarning: TypeError: server.listeners is not a function
    at Server.attach (/Users/brian/code/server/node_modules/_engine.io@3.5.0@engine.io/lib/server.js:458:26)
    at Function.attach (/Users/brian/code/server/node_modules/_engine.io@3.5.0@engine.io/lib/engine.io.js:124:10)
    at Server.initEngine (/Users/brian/code/server/node_modules/_socket.io@2.4.1@socket.io/lib/index.js:310:21)
    at /Users/brian/code/server/node_modules/_socket.io@2.4.1@socket.io/lib/index.js:295:10
    at Encoder.encode (/Users/brian/code/server/node_modules/_socket.io-parser@3.4.1@socket.io-parser/index.js:135:5)
    at Server.listen.Server.attach (/Users/brian/code/server/node_modules/_socket.io@2.4.1@socket.io/lib/index.js:290:16)
    at new Server (/Users/brian/code/server/node_modules/_socket.io@2.4.1@socket.io/lib/index.js:59:17)
    at Server (/Users/brian/code/server/node_modules/_socket.io@2.4.1@socket.io/lib/index.js:44:41)
    at IoAdapter.createIOServer (/Users/brian/code/server/node_modules/_@nestjs_platform-socket.io@7.6.15@@nestjs/platform-socket.io/adapters/io-adapter.js:25:20)
    at IoAdapter.create (/Users/brian/code/server/node_modules/_@nestjs_platform-socket.io@7.6.15@@nestjs/platform-socket.io/adapters/io-adapter.js:21:24)
(node:98373) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 2)
(node:98373) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code
```

## Input Code

```ts
// websocket.service.ts

import {
  MessageBody,
  SubscribeMessage,
  WebSocketGateway,
  WebSocketServer,
  WsResponse,
} from '@nestjs/websockets';
import { from, Observable } from 'rxjs';
import { map } from 'rxjs/operators';
import { Server } from 'socket.io';

@WebSocketGateway()
export class WebsocketService {
  @WebSocketServer()
  server: Server;

  @SubscribeMessage('events')
  findAll(@MessageBody() data: any): Observable<WsResponse<number>> {
    return from([1, 2, 3]).pipe(map(item => ({ event: 'events', data: item })));
  }

  @SubscribeMessage('identity')
  async identity(@MessageBody() data: number): Promise<number> {
    return data;
  }
}
```

```ts
// websocket.module.ts

import {Module, HttpModule} from '@nestjs/common';
import {WebsocketService} from './websocket.service';

@Module({
  imports: [
    HttpModule
  ],
  providers: [WebsocketService]
})
export class WebsocketModule {}
```

```ts
// main.ts

import { NestFactory } from '@nestjs/core';
import { AppModule } from './app.module';
import {NestExpressApplication} from '@nestjs/platform-express';
import {IoAdapter} from '@nestjs/platform-socket.io';
async function bootstrap() {
  const app = await NestFactory.create<NestExpressApplication>(AppModule);
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  app.useWebSocketAdapter(new IoAdapter(app));
  await app.listen(3000);
}
bootstrap();
```

**If I specify a port in the decorator: `@WebSocketGateway()`, this problem will not occur, but if I do not specify a port, an error will be reported**

## Expected behavior

I wonder what caused the error
## Environment

<pre><code>
Nest version: 7.5.1

 
For Tooling issues:
- Node version: v13.12.0  
- Platform:  MacBook Pro version: 10.15.7

Others:

</code></pre>

**Comment by jmcdo29** [comment]
Please provide a minimum reproduction repository.

**Comment by warriorBrian** [comment]
@jmcdo29 I expect it to work

**Comment by jmcdo29** [comment]
@warriorBrian and I expect a [minimum reproduction](http://minimum-reproduction.wtf) to showcase the issue you're facing.

**Comment by warriorBrian** [comment]
@jmcdo29 I'm replicating the problem with this repository
https://github.com/nestjs/nest/tree/master/sample/02-gateways/src

**Comment by warriorBrian** [comment]
@jmcdo29 I delete node_modules, and the error will not appear again if I download it again. I would like to know what may cause this problem.

**Comment by jmcdo29** [comment]
Most likely having `socket.io`v3 or v4 installed instead of `socket.io` v2. v3 and v4 are incompatible with v2 and there's another issue where this is tracked. Nest v8 will solve this as well, by supporting socket.io v4 which is compatible with socket.io v3

**Comment by warriorBrian** [comment]
@jmcdo29 Thank you very much 😄

**Comment by jmcdo29** [comment]
Without a minimum reproduction, no. Please open a new issue with a [minimum reprroduction](https://minimum-reproduction.wtf)

**Comment by Tinadim** [comment]
Yeah, @jmcdo29 I know what a minimum reproduction is. I was just asking in case anyone had reported it. I'm using nest@8 and Socket.io@4, which should be supported. But looks like `engine.io` is expecting the nest application (or the http server inside of it) to have a property called `listeners`, and looks like it doesn't. But I will open a new issue.

**Comment by Tinadim** [comment]
Ok, I think I have a clue of what the issue is @jmcdo29. Check this screenshot I took from Engine.io:

![Screenshot from 2022-03-02 14-56-34](https://user-images.githubusercontent.com/7933739/156420515-19ab7772-197f-4f75-a6cf-0fc379c00557.png)

I have the feeling that the server passed to socket.io (and the underlying engine.io) should be nestApp.httpServer, not nestApp itself, like this second screenshot shows:

![Screenshot from 2022-03-02 14-57-33](https://user-images.githubusercontent.com/7933739/156420643-389d28f9-7d38-4fc3-bd96-570e61766367.png)

I'm still working on a reproduction, will update this later.

**Comment by Tinadim** [comment]
Actually, nevermind. I found how to fix it:
![Screenshot from 2022-03-02 15-02-00](https://user-images.githubusercontent.com/7933739/156421129-afa8e064-7fd9-4b89-95b2-b0b97742c4d6.png)

This is the constructor of my adapter. Previously, I was passing the nest app directly to `super`. But looks like if I pass the underlying http server, it works. I got a bit confused because the name of the parameter reads `appOrHttpServer`:

![Screenshot from 2022-03-02 15-03-02](https://user-images.githubusercontent.com/7933739/156421281-6418b984-7c9b-42af-898e-107328a1249e.png)
